File: main/utopian-tree/utopian_tree.rb
def height_after_growth(cycles, height = 1, season = :spring)
  raise 'need cycles nonnegative, height positive' if cycles < 0 || height < 1

  return height if cycles == 0

  case season
  when :spring then height_after_growth(cycles - 1, height * 2, :summer)
  when :summer then height_after_growth(cycles - 1, height + 1, :spring)
  else              raise 'only :summer and :spring seasons are recognized'
  end
end

gets.to_i.times do
  puts height_after_growth(gets.to_i)
end

File: main/utopian-tree/utopian-tree.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Carp qw(croak);

use constant SPRING => 0;
use constant SUMMER => 1;

sub height_after_growth {
    (my $cycles = shift) >= 0 or croak 'cycles must be nonnegative';
    (my $height = shift // 1) > 0 or croak 'height must be positive';
    my $season = shift // SPRING;

    if ($cycles == 0) { return $height }

    if ($season == SPRING) {
        return height_after_growth($cycles - 1, $height * 2, SUMMER);
    } elsif ($season == SUMMER) {
        return height_after_growth($cycles - 1, $height + 1, SPRING);
    } else {
        croak 'unrecognized season';
    }
}

for (1..<>) {
    my $cycles = <>;
    print height_after_growth($cycles), "\n";
}
